drivers/openstack/client.go

```diff
@@ -48,44 +52,124 @@ type Client interface {
 	GetFloatingIPs(d *Driver) ([]FloatingIP, error)
 	GetFloatingIPPoolID(d *Driver) (string, error)
 	GetInstancePortID(d *Driver) (string, error)
+	VolumeCreate(d *Driver) (string, error)
+	WaitForVolumeStatus(d *Driver, status string) error
+	VolumeAttach(d *Driver) (string, error)
 }
 
 type GenericClient struct {
-	Provider *gophercloud.ProviderClient
-	Compute  *gophercloud.ServiceClient
-	Network  *gophercloud.ServiceClient
+	Provider     *gophercloud.ProviderClient
+	Compute      *gophercloud.ServiceClient
+	Network      *gophercloud.ServiceClient
+	BlockStorage *gophercloud.ServiceClient
 }
 
 func (c *GenericClient) CreateInstance(d *Driver) (string, error) {
-	serverOpts := servers.CreateOpts{
+
+	var serverOpts servers.CreateOptsBuilder
+
+	var networks []servers.Network
+	if d.NetworkId != "" {
+		networks = []servers.Network{
+			{
+				UUID: d.NetworkId,
+			},
+		}
+	}
+
+	serverOpts = &servers.CreateOpts{
 		Name:             d.MachineName,
 		FlavorRef:        d.FlavorId,
 		ImageRef:         d.ImageId,
 		UserData:         d.UserData,
+		Networks:         networks,
 		SecurityGroups:   d.SecurityGroups,
 		AvailabilityZone: d.AvailabilityZone,
 		ConfigDrive:      &d.ConfigDrive,
 	}
-	if d.NetworkId != "" {
-		serverOpts.Networks = []servers.Network{
-			{
-				UUID: d.NetworkId,
-			},
-		}
+
+	serverOpts = &keypairs.CreateOptsExt{
+		CreateOptsBuilder: serverOpts,
+		KeyName:           d.KeyPairName,
 	}
 
 	log.Info("Creating machine...")
 
-	server, err := servers.Create(c.Compute, keypairs.CreateOptsExt{
-		CreateOptsBuilder: serverOpts,
-		KeyName:           d.KeyPairName,
-	}).Extract()
+	var server *servers.Server
+	var err error
+
+	if d.BootFromVolume {
+		blockDevices := []bootfromvolume.BlockDevice{
+			bootfromvolume.BlockDevice{
+				BootIndex:           0,
+				DeleteOnTermination: true,
+				DestinationType:     bootfromvolume.DestinationVolume,
+				SourceType:          bootfromvolume.SourceImage,
+				UUID:                d.ImageId,
+				VolumeType:          d.VolumeType,
```

Comment:
```suggestion

```

Reply:
In the gophercloud docs, `This requires Compute API microversion 2.67 or later`. Tested on an OpenStack setup with compute min microversion 2.1 and max microversion 2.79 (and for what it's worth, block min 3.0 and max 3.59).

Reply:
Hello, this is something I would really need, I've a use case where ROOT volume need to be on a "ssd" disk pool.
(ex. a node-template for k8s master where etcd requires fast I/O.)
As you mention this requires specific `Compute API microversion 2.67 or later`, but if you leave it empty it doesn't break and choose you default volume_type/pool.

Reply:
Ahhh, got it, I think I misread some code and thought this wasn't necessary for booting. Still learning, thanks for the clarification!

What's the best way to force my rancher-machine to use this version of the compute API? My setup seems to support it but I guess it's not using that API version for some reason?

Reply:
Do you have access to an Openstack Stein version or later ?

Reply:
I have access to a Train cluster I believe. If you make a new test build with your latest compute version change I can give it a try pretty quick

Reply:
I think it's ok now. I tested it against a lower version and I saw the API rejecting the instance creation and I saw specifically the client had the 2.67 microversion header.

Reply:
@thpham do you have a new build at your `thpham/rancher:dev` tag?

Reply:
here it is, the new image (based on rancher:master) with the current changes of this PR: `tpham/rancher:8acfe59c2`

Reply:
Sweet, giving it a go now.

Reply:
Worked for me! Volume type was set correctly, no errors thrown, VM provisioned properly.